feat(HMS-1550): add notifications service integration

amirfefer opened the pull request:
This PR adds notification support, and sends a  message via kafka once launch process is finished.

We currently support two event types under Image builder application:
- Launch completed
- Launch failed

**EDIT:
In order to test locally, follow the README.notifications** 

In order to test it under stage env:
1. Go to Notifications -> Integreation 
2.  create a new webhook integration, you can use https://webhook.site/ tool that will provide a unique webhook URL
3. Go to Notifications -> RHEL
4. Create a new behaiver group, assign one of the above event types (i.e `Launch failed`) and the integration you have created.

Please note that only webhooks work ATM,  email templates will be added later.

amirfefer commented:
Turned into a draft.

amirfefer commented:
lcaol webhook triggering:

https://user-images.githubusercontent.com/11807069/234266180-e511ae0a-f39b-44b2-90d2-22e0d138edf1.mov

Ready for review :)

lzap commented on internal/kafka/notification_msg.go:
```diff
@@ -0,0 +1,76 @@
+)
+
+type NotificationEvent struct {
+	Payload interface{} `json:"payload"`
```
Using empty interface is almost always a red flag in Go. We should stick with types as hard as possible.

In this case, use [RawMessage](https://pkg.go.dev/encoding/json#RawMessage) type instead. You can implement MarshalJSON method for NotificationEvent and NotificationError/NotificationSuccess types and use it efficiently while taking advantage of types.

Btw these types should go into a new package named `notifications` or `ntf` maybe, then you can drop all the Notification prefix/suffis. I mentioned that above that there should be also an interface in the newly created package that hides the logic.

Here is a working example on how to work with RawMessage: https://go.dev/play/p/LU-w97uHF-5

lzap commented on scripts/notifications/helpers.py:
```diff
@@ -0,0 +1,251 @@
+
+    print(r.status_code)
+
+    return None
```
Is this your code or taken from somewhere?

amirfefer replied:
this is taken from the `notification-backend` repo, but it's not needed anymore

lzap reviewed (changes requested):
Apologies for late review.

It looks great, I have two major requests.

First, the notification should live in its own package and there should be an interface that other parts of the application can easily use. They should not be exposed to implementation "details" or the fact they actually call Kafka - this should be all hidden. The same package should define "model" types that are then serialized to JSON.

Kafka should not be the only implementation, you should also provide a "dummy" implementation that does "nothing". This should be used in case the feature is not enabled. We have multiple places where we use this approach, for example DAO, [cache](https://github.com/RHEnVision/provisioning-backend/tree/main/internal/cache) or hyperscaler clients.

Second, this should be a configurable feature, so please add a new value to `config.go` either under APP_ or create new group NOTIFICATIONS_ with "enabled" boolean flag. The default value should be - disabled. We do not want this in development mode, on stage and prod this should be enabled. You can achieve this in `clowdapp.yml` easily.

amirfefer commented on internal/kafka/notification_msg.go:
```diff
@@ -0,0 +1,76 @@
+)
+
+const (
+	application                  = "image-builder"
```
no limitations, there's an image-builder application, and the launch wizard from the UI/UX  perspective is part of the image-builder application.

amirfefer commented:
@lzap, thanks for your review!  I addressed your comments, also there's a PR for adding the notifications profile in compose config - https://github.com/RHEnVision/provisioning-compose/pull/3

lzap commented on internal/notifications/noop.go:
```diff
@@ -0,0 +1,26 @@
+
+type noopNotificationClient struct{}
+
+// var _ NotificationClient = &noopNotificationClient{}
```
You can keep this, the reason is that the compiler will show you more readable error when you miss a method in the implementation. Or you can delete if you don’t like it.

lzap commented on internal/notifications/notifications.go:
```diff
@@ -0,0 +1,89 @@
+	rDao := dao.GetReservationDao(ctx)
+	reservation, err := rDao.GetById(ctx, reservationId)
+	if err != nil {
+		logger.Warn().Err(err).Msg("unable to find reservation by id")
```
Nitpick:

* Log messages start with capital letter
* Go errors must be all lowercase (linter enforces it)

I know, if I could ever change this decision, I would use lowercase everywhere.

lzap commented on internal/notifications/notifications.go:
```diff
@@ -0,0 +1,89 @@
+		NotificationInstancesEvents[i] = kafka.NotificationEvent{Payload: marshalInstance}
+	}
+
+	notificationMsg, err := kafka.NotificationMessage{Context: reservation, EventType: kafka.NotificationSuccessEventType, Events: NotificationInstancesEvents}.GenericMessage(ctx)
```
Quite long line, I suggest to break into multiple lines.

lzap commented on internal/notifications/interface.go:
```diff
@@ -0,0 +1,12 @@
+type NotificationClient interface {
+	SuccessfulLaunch(ctx context.Context, reservationId int64)
+	FailedLaunch(ctx context.Context, reservationId int64, jobError string)
+}
```
Nice, so the API is designed in a way that it ignores errors. It is fine.

lzap commented on internal/notifications/notifications.go:
```diff
@@ -0,0 +1,89 @@
+	err = kafka.Send(ctx, &notificationMsg)
+	if err != nil {
+		logger.Warn().Err(err).Msg("Unable to send notification message via kafka")
+	}
```
This must be error log, because warnings do not propagate into Sentry. We want to know if Kafka is down via Sentry.

lzap commented on internal/notifications/notifications.go:
```diff
@@ -0,0 +1,89 @@
+	err = kafka.Send(ctx, &notificationMsg)
+	if err != nil {
+		logger.Warn().Err(err).Msg("Unable to send notification message via kafka")
+	}
```
This must be error log, because warnings do not propagate into Sentry. We want to know if Kafka is down via Sentry.

lzap reviewed (changes requested):
Absolutely perfect!

I dropped few small nitpicks and one logging/sentry log level problem.

amirfefer commented on internal/jobs/launch_instance_gcp.go:
```diff
@@ -50,15 +51,21 @@ func HandleLaunchInstanceGCP(ctx context.Context, job *worker.Job) {
+		nc.SuccessfulLaunch(ctx, args.ReservationID)
+	}
 	finishJob(ctx, args.ReservationID, jobErr)
 }
```
> The function finishWithSuccess is meant to be generic. Currently we only have jobs which launch instances but we might create jobs that do different stuff. I think it might be beneficial to put this code into individual jobs as extra DoSendNotification functions.

finishWithSuccess / finishWithError aren't meant to be generic?

lzap replied:
True, what you have is better, yeah. They actually take reservationID meaning that they should be only for reservations, but I don’t like this design myself. It is unclear, I could not figure out a better solution.

amirfefer commented:
@lzap, I've added  configMap for seeding data in ephemera, we can test triggered events via `api/notifications/v2.0/notifications/events` endpoint

lzap reviewed (changes requested):
It is not starting up service for workers.

lzap commented:
/retest

lzap commented:
hit fatal error: 'oc process' command failed: The Template "notifications-backend" is invalid: template.parameters[26]: Required value: template.parameters[26]: parameter FLOORIST_SCHEDULE is required and must be specified

So I contacted the ntf team and they are currently transitioning to Floorist Operator and have an issue for ephemeral. They are working on it, so we will have to wait. Fix is planned for later today.

MikelAlejoBR commented:
/retest

lzap commented:
/retest

lzap commented on internal/notifications/notifications.go:
```diff
@@ -0,0 +1,92 @@
+
+func init() {
+	GetNotificationClient = getNoopNotificationClient
+}
```
I think the problem is here.

You do not need to use init function, in fact, this Go feature should be avoided if possible. It works for database or kafka beacuse implementations are in different packages, but here you have both implementations in the same package.

Can you simply remove init and do this:

func Initialize(ctx context.Context) {
  if config.Application.Notifications.Enabled {
    zerolog.Ctx(ctx).Debug().Msg("Initialized kafka notification client")
    GetNotificationClient = getNotificationClient
  } else {
    zerolog.Ctx(ctx).Debug().Msg("Initialized noop notification client")
    GetNotificationClient = getNoopClient
  }	
}

lzap commented:
Found the problem, Kafka is not configured for workers, they simply do not use Kafka up until this point. Add this bit into `worker.go` as well:

```
	// initialize platform kafka
	if config.Kafka.Enabled {
		err = kafka.InitializeKafkaBroker(ctx)
		if err != nil {
			logger.Fatal().Err(err).Msg("Unable to initialize the platform kafka")
		}
		if config.Application.Notifications.Enabled {
			notifications.Initialize(ctx)
		}
	}
```

lzap reviewed (approved):
Makes sense. Re-testing.

lzap reviewed (changes requested):
Problem with initialization.

lzap commented:
I see `AttributeError: 'Namespace' object has no attribute 'default_api'` failures, probably unrelated? @akhil-jha 

/retest

lzap commented:
Belive it or not, it does not work. I don’t see notification initialized for worker:

```
7:21PM TRC Cloudwatch not enabled, enabling stdout binary=worker hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM INF Clowder environment: env-ephemeral-5qhl4j hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM INF Clowder database hostname: provisioning-backend-db.ephemeral-5qhl4j.svc hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM INF Clowder kafka brokers: env-ephemeral-5qhl4j-5af11a49-kafka-bootstrap.ephemeral-5qhl4j.svc hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM INF Clowder logging type: null, region: , group:  hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM INF Configuration: {App:{Port:8000 InstancePrefix: Notifications:{Enabled:true} Cache:{Type:redis Expiration:1h0m0s Redis:{Host:provisioning-backend-redis.ephemeral-5qhl4j.svc Port:6379 User: Password: DB:0} Memory:{CleanupInterval:5m0s}}} Database:{Host:provisioning-backend-db.ephemeral-5qhl4j.svc Port:5432 Name:provisioning User:nxgFB1J8jr2CbfDu Password:**** SeedScript: MinConn:2 MaxConn:50 MaxIdleTime:15m0s MaxLifetime:2h0m0s LogLevel:info} Logging:{Level:trace Stdout:true MaxField:0} Telemetry:{Enabled:true Jaeger:{Enabled:false Endpoint:http://localhost:14268/api/traces} Logger:{Enabled:true}} Cloudwatch:{Enabled:false Region: Key:**** Secret:**** Session:**** Group: Stream:} AWS:{Key:**** Secret:**** Session:**** DefaultRegion:us-east-1 Logging:false} Azure:{TenantID:520cf09d-78ff-44ed-a731-abd623e73b09 ClientID:**** ClientSecret:**** ClientPrincipalID: ClientPrincipalName:RH HCC DefaultRegion:eastus SubscriptionID:} GCP:{ProjectID: JSON:**** DefaultZone:us-east4} Prometheus:{Port:9000 Path:/metrics} RestEndpoints:{ImageBuilder:{URL:http://image-builder-service.ephemeral-5qhl4j.svc:8000/api/image-builder/v1 Username: Password:**** Proxy:{URL:}} Sources:{URL:http://sources-api-svc.ephemeral-5qhl4j.svc:8000/api/sources/v3.1 Username: Password:**** Proxy:{URL:}} TraceData:true} Worker:{Queue:redis PollInterval:5s Concurrency:33 Timeout:30m0s} Unleash:{Enabled:true Environment: Prefix:provisioning URL:http://env-ephemeral-5qhl4j-featureflags.ephemeral-5qhl4j.svc:4242/api Token:****} Sentry:{Dsn:} Kafka:{Enabled:true Brokers:[env-ephemeral-5qhl4j-5af11a49-kafka-bootstrap.ephemeral-5qhl4j.svc:9092] AuthType: CACert: SASL:{Username:**** Password:**** SaslMechanism: SecurityProtocol:}}} hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM INF Worker starting hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM INF Initializing redis application cache cache=true hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Initializing database connection hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Initializing 'redis' job queue hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Registering job queue handlers and interfaces hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Starting dequeue loop hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM INF Starting Redis dequeuer with 33 polling goroutines hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 151ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 1208ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 0ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 302ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 604ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 906ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 453ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 1057ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 755ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 2114ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 3020ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 3926ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 4379ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 1359ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 3473ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 3624ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 4077ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 2265ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 1510ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 4228ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 2416ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 3171ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 3775ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 1661ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 2567ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 1812ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 3322ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 4530ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 1963ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 2718ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 4681ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 2869ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Worker start delay 4832ms hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM INF Starting new instance on port 8000 with prometheus on 9000 hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:21PM DBG Job queue metric delay 27639ms background=true hostname=provisioning-backend-worker-5d974f965c-pg4vk version=30f6
7:29PM INF Dequeued job from Redis hostname=provisioning-backend-worker-5d974f965c-pg4vk job_args={"Fail":false,"ReservationID":1,"Sleep":0} job_id=42b10f56-ef6b-4efe-811a-1a72ece44e08 job_type=no_operation version=30f6
7:29PM DBG Reservation status change: 'No operation started' account_id=1 account_number=12345 hostname=provisioning-backend-worker-5d974f965c-pg4vk job_id=42b10f56-ef6b-4efe-811a-1a72ece44e08 org_id=12345 reservation_id=1 step=true version=30f6
7:29PM INF No operation finished account_id=1 account_number=12345 hostname=provisioning-backend-worker-5d974f965c-pg4vk job_id=42b10f56-ef6b-4efe-811a-1a72ece44e08 org_id=12345 reservation_id=1 version=30f6
7:29PM DBG Reservation status change: 'No operation finished' account_id=1 account_number=12345 hostname=provisioning-backend-worker-5d974f965c-pg4vk job_id=42b10f56-ef6b-4efe-811a-1a72ece44e08 org_id=12345 reservation_id=1 step=true version=30f6
7:29PM TRC Increased step number by: 1 account_id=1 account_number=12345 hostname=provisioning-backend-worker-5d974f965c-pg4vk job_id=42b10f56-ef6b-4efe-811a-1a72ece44e08 org_id=12345 reservation_id=1 step=true version=30f6
7:29PM DBG Job step: 1/1 account_id=1 account_number=12345 hostname=provisioning-backend-worker-5d974f965c-pg4vk job_id=42b10f56-ef6b-4efe-811a-1a72ece44e08 org_id=12345 reservation_id=1 version=30f6
7:29PM INF All jobs executed, marking job as success account_id=1 account_number=12345 hostname=provisioning-backend-worker-5d974f965c-pg4vk job_id=42b10f56-ef6b-4efe-811a-1a72ece44e08 org_id=12345 reservation_id=1 version=30f6
7:29PM INF Dequeued job from Redis hostname=provisioning-backend-worker-5d974f965c-pg4vk job_args={"Fail":false,"ReservationID":2,"Sleep":0} job_id=6e84437c-31bb-4579-bfc8-22777070b4b0 job_type=no_operation version=30f6
7:29PM DBG Reservation status change: 'No operation started' account_id=1 account_number=12345 hostname=provisioning-backend-worker-5d974f965c-pg4vk job_id=6e84437c-31bb-4579-bfc8-22777070b4b0 org_id=12345 reservation_id=2 step=true version=30f6
7:29PM INF No operation finished account_id=1 account_number=12345 hostname=provisioning-backend-worker-5d974f965c-pg4vk job_id=6e84437c-31bb-4579-bfc8-22777070b4b0 org_id=12345 reservation_id=2 version=30f6
7:29PM DBG Reservation status change: 'No operation finished' account_id=1 account_number=12345 hostname=provisioning-backend-worker-5d974f965c-pg4vk job_id=6e84437c-31bb-4579-bfc8-22777070b4b0 org_id=12345 reservation_id=2 step=true version=30f6
7:29PM TRC Increased step number by: 1 account_id=1 account_number=12345 hostname=provisioning-backend-worker-5d974f965c-pg4vk job_id=6e84437c-31bb-4579-bfc8-22777070b4b0 org_id=12345 reservation_id=2 step=true version=30f6
7:29PM DBG Job step: 1/1 account_id=1 account_number=12345 hostname=provisioning-backend-worker-5d974f965c-pg4vk job_id=6e84437c-31bb-4579-bfc8-22777070b4b0 org_id=12345 reservation_id=2 version=30f6
7:29PM INF All jobs executed, marking job as success account_id=1 account_number=12345 hostname=provisioning-backend-worker-5d974f965c-pg4vk job_id=6e84437c-31bb-4579-bfc8-22777070b4b0 org_id=12345 reservation_id=2 version=30f6
```

lzap commented on cmd/pbackend/api.go:
```diff
@@ -64,12 +65,13 @@ func api() {
 		if err != nil {
 			logger.Fatal().Err(err).Msg("Unable to initialize the platform kafka")
 		}
+		notifications.Initialize(ctx)
```
```suggestion
        if config.App.Notifications {
		        notifications.Initialize(ctx)
		}
```

lzap commented on cmd/pbackend/statuser.go:
```diff
@@ -278,6 +279,9 @@ func statuser() {
 
+	// initialize notifications
+	notifications.Initialize(ctx)
+
```
Can you also make this the same as the other two?

lzap commented on cmd/pbackend/worker.go:
```diff
@@ -52,6 +54,15 @@ func worker() {
+			logger.Fatal().Err(err).Msg("Unable to initialize the platform kafka")
+		}
+		notifications.Initialize(ctx)
+	}
```
Ditto.

lzap reviewed (changes requested):
I am not sure why Kafka is not starting, config is set. Will take a look locally.

akhil-jha commented:
> I see `AttributeError: 'Namespace' object has no attribute 'default_api'` failures, probably unrelated? @akhil-jha
> 
> /retest

Something to be fixed on the qe test level. Working on that

lzap reviewed (approved):
It works, I will do a small followup after this is merged, don’t worry about my last review.

@akhil-jha to test this feature, create a noop reservation and then hit notifications backend:

```
{
  "data": [
    {
      "id": "11cffb32-e4ab-4c15-81d0-b9819a23bf55",
      "created": "2023-06-23T06:30:59.262905",
      "bundle": "Red Hat Enterprise Linux",
      "application": "Image Builder",
      "event_type": "Launch Success",
      "payload": null,
      "actions": []
    },
    {
      "id": "69690183-1e1e-4bdc-acbe-1757fe3fc6df",
      "created": "2023-06-23T06:30:59.058037",
      "bundle": "Red Hat Enterprise Linux",
      "application": "Image Builder",
      "event_type": "Launch Success",
      "payload": null,
      "actions": []
    }
  ],
  "links": {
    "last": "/api/notifications/v1.0/notifications/events?limit=20&offset=0",
    "first": "/api/notifications/v1.0/notifications/events?limit=20&offset=0"
  },
  "meta": {
    "count": 2
  }
}
```

Can we merge and do a new IQE test as a followup? Not a critical feature, it is designed in a way that if anything fails, notification will simply not be delivered.

akhil-jha commented:
I'd say let the pr_check pass :) I have a fix ready. I'll add a retest and merge this.

akhil-jha commented:
/retest

akhil-jha commented:
Failures because of source availability for superkey and gcp.  Not related failures.

akhil-jha reviewed (approved):
QE test will follow.

lzap commented:
Thanks all.

lzap commented:
Follow-up: https://github.com/RHEnVision/provisioning-backend/pull/576
